### PR TITLE
Make reflection.h types C++20 compatible.

### PR DESCRIPTION
**Pull request by BertusGreeff**
Fix for issue Bond build failed with error C5208 in MSVC #1027

### CODE REVIEW

File: compiler/src/Language/Bond/Codegen/Cpp/Reflection_h.hs [hunk 1]
```diff
@@ -101,21 +101,23 @@ reflection_h export_attribute cpp file imports declarations = ("_reflection.h",
...
                 #{fieldOrdinal},
                 #{CPP.modifierTag f},
                 #{className},
                 #{cppType fieldType},
                 &#{className}::#{fieldName},
                 &s_#{fieldName}_metadata
-            > {}  #{fieldName};
+            > {};
+
+            struct #{fieldName} #{fieldName};
```
Comment (chwarr) [inline compiler/src/Language/Bond/Codegen/Cpp/Reflection_h.hs]: I don't think we need this member variable.

**Comment by bhardwajs** [comment]
@BertusGreeff - Are you still working on this PR? I am working on removing dependency on MSVC 2013 and upgraded my MSVC to 16.5.4 and hit these issues. I would rather merge your changes than redo those :)

**Comment by BertusGreeff** [comment]
Yes, I’m still working on this.  It’s a part time job,  Was going to look at it tomorrow again.

**Comment by bhardwajs** [comment]
> Yes, I’m still working on this. It’s a part time job, Was going to look at it tomorrow again.

Great. Thank you! Looking forward to the update :)

File: compiler/src/Language/Bond/Codegen/Cpp/Reflection_h.hs [hunk 2]
```diff
@@ -106,16 +106,25 @@ reflection_h export_attribute cpp file imports declarations = ("_reflection.h",
...
 
-        fieldTemplates = F.foldMap $ \ f@Field {..} -> [lt|
+        -- fieldTemplateReservedNames are static variables names used in ::bond::reflection::FieldTemplate<>
+        fieldTemplateReservedNames = ["metadata", "field", "id"]
+        
+        fieldNames = map (\f -> fieldName f) structFields
+
+        fieldTemplateStructReservedNames = fieldTemplateReservedNames ++ fieldNames
+
+        uniqueFieldTemplateStructNames = uniqueNames (map (\n -> "_" ++ n) fieldNames) fieldTemplateStructReservedNames
```
Comment (ara-ayvazyan) [inline compiler/src/Language/Bond/Codegen/Cpp/Reflection_h.hs:116]: We should also avoid using [reserved identifiers](https://en.cppreference.com/w/cpp/language/identifiers).
> the identifiers with a double underscore anywhere are reserved;
> the identifiers that begin with an underscore followed by an uppercase letter are reserved;
> the identifiers that begin with an underscore are reserved in the global namespace.

Comment (BertusGreeff) [reply compiler/src/Language/Bond/Codegen/Cpp/Reflection_h.hs]: Addressed in latest revision.

**Pull request merged**